src/main/scala/org/constellation/gossip/bisect/Bisect.scala

```diff
@@ -0,0 +1,38 @@
+package org.constellation.gossip.bisect
+
+import cats.data.OptionT
+import cats.effect.Async
+import cats.implicits._
+import io.chrisdavenport.log4cats.SelfAwareStructuredLogger
+import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+
+class Bisect[F[_]: Async] {
+  private val logger: SelfAwareStructuredLogger[F] = Slf4jLogger.getLogger[F]
+
+  def runA[A](
+    predicate: A => F[Boolean],
+    initSeq: Seq[A]
+  ): F[Option[A]] = {
+    def go(seq: Seq[A], acc: OptionT[F, A]): OptionT[F, A] = seq match {
```

Comment:
what do you think about @tailrec optimization? it may work here 🤔

Reply:
I tried that last time, and I couldn't come up with a solution. I could give it another shot, but I don't promise any results 😅

Reply:
I don't know if this helps but you could try using tailrecM ;)